Source/WebCore/dom/InternalObserverForEach.cpp

```diff
@@ -0,0 +1,137 @@
+*    notice, this list of conditions and the following disclaimer in the
+*    documentation and/or other materials provided with the distribution.
+*
+* THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+* AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+* THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+* PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+* BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+* CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+* SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+* INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+* CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+* ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+* THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+#include "config.h"
+#include "InternalObserverForEach.h"
+
+#include "AbortSignal.h"
+#include "InternalObserver.h"
+#include "JSDOMPromiseDeferred.h"
+#include "Observable.h"
+#include "ScriptExecutionContext.h"
+#include "SubscribeOptions.h"
+#include "Subscriber.h"
+#include "SubscriberCallback.h"
+#include "VisitorCallback.h"
+#include <JavaScriptCore/JSCJSValueInlines.h>
+
+namespace WebCore {
+
+class InternalObserverForEach final : public InternalObserver {
+public:
+    static Ref<InternalObserverForEach> create(ScriptExecutionContext& context, Ref<VisitorCallback>&& callback, Ref<AbortSignal>&& signal, Ref<DeferredPromise>&& promise)
+    {
+        Ref internalObserver = adoptRef(*new InternalObserverForEach(context, WTFMove(callback), WTFMove(signal), WTFMove(promise)));
+        internalObserver->suspendIfNeeded();
+        return internalObserver;
+    }
+
+private:
+    void next(JSC::JSValue value) final
+    {
+        auto* globalObject = protectedScriptExecutionContext()->globalObject();
+        ASSERT(globalObject);
+
+        Ref vm = globalObject->vm();
+
+        {
+            JSC::JSLockHolder lock(vm);
+
+            // The exception is not reported, instead it is forwarded to the
+            // abort signal and promise rejection.
+            // As such, VisitorCallback `[RethrowsException]` and here a
+            // catch scope is declared so the error can be passed to any promise
+            // rejection handlers and abort handlers.
+            auto scope = DECLARE_CATCH_SCOPE(vm);
+
+            Ref { m_callback }->handleEvent(value, m_idx++);
+
+            JSC::Exception* exception = scope.exception();
+            if (UNLIKELY(exception)) {
+                scope.clearException();
+                auto value = exception->value();
+                Ref { m_promise }->reject<IDLAny>(value);
+                Ref { m_signal }->signalAbort(value);
+            }
+        }
+    }
+
+    void error(JSC::JSValue value) final
+    {
+        Ref { m_promise }->reject<IDLAny>(value);
```

Comment:
Our rules of thumb in WebKit is that if you have to call `Ref { m_promise }` more than once, then your should be adding a protected getter:
```
Ref<DeferredPromise> protectedPromise() const { return m_promise; }
```
then call `protectedPromise()->reject()` here. Same comment applies elsewhere.

Reply:
Given we let this go through in https://github.com/WebKit/WebKit/pull/35261/files#diff-42a899eeb66d8a4163f94ff20bdc4d0cfeaf4fb2c7a3e1a3ba69906c29f9e741R65 are you okay for me to action this as part of https://bugs.webkit.org/show_bug.cgi?id=282074 for a broader cleanup of the Observable code?

Reply:
I'd rather we correct this PR before landing.

Reply:
Sounds good to me, Ill do this right away!

Reply:
I've pushed that change, please let me know if I've put those protected getter's in the right place (in terms of order). I want to learn the ordering convention.